### PR TITLE
feat: add autoHeaderHeight option for multi-line column headers

### LINKED ISSUE #1272: Support header height auto-sizing
### Clear and concise description of the problem

I want the header row to size itself to whatever content is actually rendered in .slick-header-column (multi-line HTML titles or DOM titles) so that column titles aren't silently clipped whenever they need more than one line.

### Suggested solution

Add an opt-in grid option, e.g. `autoHeaderHeight`.

### Alternative

_No response_

### Additional context

_No response_

### Validations

- [x] Follow our [Code of Conduct](https://github.com/6pac/SlickGrid/blob/master/CODE_OF_CONDUCT.md)
- [x] Read the [Wikis](https://github.com/6pac/SlickGrid/wiki).
- [x] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### PR DESCRIPTION
**Pull request by jahanbakhsh18**
## Add `autoHeaderHeight` option for multi-line column headers

This PR adds a new `autoHeaderHeight` grid option that allows column headers to auto-size based on their content.

Closes #1272 

### Features
- Disabled by default to preserve existing behavior.
- Recalculate the header height:
  - after column headers are created
  - after a column resize is completed
  - when `autoHeaderHeight` is enabled or disabled through `setOptions`
- Supports frozen columns by keeping left and right header panes aligned.
- Adds a dedicated example and Cypress test coverage.

The grid core handles measuring and synchronizing the required header height. The calculated dimensions are passed to the themes through CSS custom properties, allowing to retain their respective header layout behavior.

### Demo
A short demo video and screenshots  are attached below.

https://github.com/user-attachments/assets/b9fab68a-8cab-4f29-87cc-ce817f213468

example-auto-header-height

### CODE REVIEW

File: src/styles/slick-alpine-theme.scss [hunk 1]
```diff
@@ -608,6 +608,31 @@
   display: block;
 }
 
+.slick-header-auto-height {
+  --alpine-header-align-items: center;
```
Comment (ghiscoding) [inline src/styles/slick-alpine-theme.scss]: it's great to have a CSS variable but I would assume that most users would it aligned to top not centered. 

You should also create a SASS variable and use the CSS variable fallback like I did everywhere in this file so that SASS users could also take advantage of this

File: src/styles/slick-alpine-theme.scss [hunk 2]
```diff
@@ -608,6 +608,31 @@
...
+  .slick-header-columns-right {
+    height: var(--slick-auto-header-height);
+    white-space: normal;
+    overflow: visible;
+  }
+
+  .slick-header-column {
+    height: calc(
+      var(--slick-auto-header-height) -
+      var(--slick-auto-header-height-extra, 0px)
```
Comment (ghiscoding) [inline src/styles/slick-alpine-theme.scss]: as mentioned above use CSS fallback to SASS variable

File: src/slick.grid.ts [hunk 3]
```diff
@@ -1939,6 +1940,98 @@ export class SlickGrid<TData = any, C extends Column<TData> = Column<TData>, O e
...
+    headers.forEach(header => {
+      header.classList.toggle('slick-header-auto-height', enabled);
+    });
+
+    if (!enabled) {
+      this._clearAutoHeaderHeightStyles(headers);
+      return;
+    }
+
+    setTimeout(() => this.recalculateHeaderHeight(), 0);
```
Comment (ghiscoding) [inline src/slick.grid.ts]: is this delay really necessary? I'm trying to avoid using delay like this in the code whenever possible

File: src/slick.grid.ts [hunk 4]
```diff
@@ -1939,6 +1940,98 @@ export class SlickGrid<TData = any, C extends Column<TData> = Column<TData>, O e
...
+      header.style.removeProperty('--slick-auto-header-height-extra');
+      header.style.height = '';
+    });
+  }
+
+  /**
+   * Applies the calculated automatic header height styles.
+   */
+  private _setAutoHeaderHeightStyles(height: number, extra: number, headers: HTMLDivElement[]) {
+    const heightValue = `${height}px`;
```
Comment (ghiscoding) [inline src/slick.grid.ts]: I think we can drop these constants and go straight to the assignment below since the CSS variables already tell us what it is

File: src/styles/slick-alpine-theme.scss (hunk 1 above)
Comment (jahanbakhsh18) [reply src/styles/slick-alpine-theme.scss]: Agreed. `autoHeaderHeight` should control the header height, not force a vertical alignment preference. I'll remove the `--alpine-header-align-items: center` override and let the theme's existing default alignment apply.
(I initially chose center because it looked better to my eye 🤗)

File: src/slick.grid.ts (hunk 3 above)
Comment (jahanbakhsh18) [reply src/slick.grid.ts]: That's a fair point. The timeout was a simple way to ensure the measurement happened after the new header layout was rendered. I agree that it's better to avoid arbitrary delays. I've now moved the recalculation to a more appropriate lifecycle stage

File: src/styles/slick-alpine-theme.scss (hunk 2 above)
Comment (jahanbakhsh18) [reply src/styles/slick-alpine-theme.scss]: I've added `$alpine-auto-header-height-extra` to the _variables.scss file, following the project's naming convention. The Alpine theme now uses this SASS variable as a fallback for the CSS custom property:

```scss
height: calc(
   var(--slick-auto-header-height) -
   var(--alpine-auto-header-height-extra, v.$alpine-auto-header-height-extra)
);
```

Why `--slick-auto-header-height` doesn't have a SASS variable: This value is dynamic and comes from JavaScript (`header.getBoundingClientRect().height`). It cannot be defined at compile time, so it must remain a CSS custom property set by the code.

Comment (ghiscoding) [reply src/styles/slick-alpine-theme.scss]: SASS is a fallback, you can still define it as an empty SASS variable, the advantage is that if the user isn't using this feature, he could provide something else via the SASS variable (like `center`) and that would still work. 

The CSS variable is the main entry point, so if it's provided or changed by JS code then it will be used and discard the SASS usage completely, so again that would work without any issue... the point is that with/without the SASS variable, the JS code would still work because CSS variable has always higher priority and SASS is always a fallback with lower priority. I actually took advantage of that when creating the new Alpine Theme which gives us a 2-in-1 approach :)

so anyway, it won't make much difference if you don't provide the SASS variable but it's just for consistency (even if it's defined as empty)

**Comment by ghiscoding** [comment]
@jahanbakhsh18 feel free to click on the the "Resolve Comment" button when you've addressed any of the comments I left.

also CCing @muendlein as well so that he's aware of the PR and can also provide feedback

**Comment by muendlein** [comment]
From my side I would also like to see an example where a column header is multiline HTML string or a DOM element.
Preferably also in a scenario with fit column autosizing.
In this context I see the use case for some finer control on per column level. One example use case is that a column with a multiline HTM string should not be cut off while other plain string column headers should not be wrapped. This is especially relevant in case a column header has many short words which can lead to excessive header height when wrapped.

**Comment by jahanbakhsh18** [comment]
Thanks @muendlein for the suggestion. I created an additional example to check the feasibility of these scenarios with the current `autoHeaderHeight` implementation.
I tested the following header types:
- A multi-line HTML string
- Another multi-line HTML string with different content styling
- A DOM element used directly as the column name
- A regular long text header

```html
...
const headerElement = document.createElement('p');
headerElement.style.margin= 0;
headerElement.innerHTML = `Customer <span style="color:red;">Information</span>`;

var columns = [
  { id: "sel", name: "#", field: "num", behavior: "select", cssClass: "cell-selection", 
    width: 30, cannotTriggerInsert: true, resizable: false, unselectable: true 
  },
  { id: "html-header", name: "Customer<br><strong style=\"color:red;\">Information</strong>", 
    field: "title", width: 100 
  },
  { id: "html-header-2", name: "Project<br><span style='font-size: 8px'>Status & Details</span>", 
    field: "title2", width: 100 
  },
  { id: "header-element", name: headerElement, field: "title3", width: 100 },
  { id: "description", name: "Very Long Header With Several Words", field: "title4", width: 150 }
];

var options = { 
  enableCellNavigation: true, enableColumnReorder: false, autoHeaderHeight: true, frozenColumn: 1 
};
...
```

html-element-long

#### Note:
The current implementation handles all of these cases without any additional changes. The header height is calculated from the rendered content, so the multi-line HTML and DOM element are included naturally in the measurement.
I also agree that finer per-column control could be useful, especially for cases where only selected columns should wrap while other columns should remain on a single line. However, that would introduce a separate per-column wrapping/layout API and is beyond the scope of the current PR, whose goal is to provide the basic infrastructure for automatic header height calculation and synchronization.

File: src/slick.grid.ts [hunk 5]
```diff
@@ -1939,6 +1945,73 @@ export class SlickGrid<TData = any, C extends Column<TData> = Column<TData>, O e
...
+    const headers = [this._headerScrollerL, this._headerScrollerR]
+      .filter((header): header is HTMLDivElement => !!header);
+
+    headers.forEach(header => {
+      header.classList.toggle('slick-header-auto-height', enabled);
+    });
+
+    if (!enabled) {
+      this._clearAutoHeaderHeightStyles(headers);
+      return;
```
Comment (ghiscoding) [inline src/slick.grid.ts]:
```suggestion
```

this return is irrelevant since this is the last line executed anyway, removing this line will make no difference, just less code ;)

**Comment by muendlein** [comment]
@jahanbakhsh18 Thank you, this is looking good!
For the example itself what happens when a autosizing mode like `IgnoreViewport` is active? Especially in combination with using the `ignoreHeaderText` option.

**Comment by ghiscoding** [comment]
Hey everyone, I asked Codex (ChatGPT 5.6) to review PR #1282, all PR comments, and the original issue #1272. This is only a second opinion for discussion, not a maintainer verdict.

## Context from issue #1272

The original issue is specifically about preventing rendered content inside `.slick-header-column` from being silently clipped. This includes:

- Multi-line HTML header titles
- DOM elements used as header titles
- Headers requiring more than two lines
- Frozen column and frozen row layouts
- Recalculation when column widths or the grid container size changes

The original issue prototype also identified a frozen-layout problem where the two header panes had different heights and the grid overflowed its container.

## Review findings

### 1. High — header columns can still clip rendered content

The PR changes `.slick-column-name` to `overflow: visible`, but `.slick-header-column` itself still has `overflow: hidden` in both the base styles and Alpine theme.

Therefore, multi-line HTML or DOM content can still be clipped by its parent header column.

Relevant rules:

- [slick.grid.scss](https://github.com/6pac/SlickGrid/blob/e6548939c8bf336548adbf29750f148e51aad30c/src/styles/slick.grid.scss#L30-L38)
- [slick.grid.scss](https://github.com/6pac/SlickGrid/blob/e6548939c8bf336548adbf29750f148e51aad30c/src/styles/slick.grid.scss#L277-L294)
- [slick-alpine-theme.scss](https://github.com/6pac/SlickGrid/blob/e6548939c8bf336548adbf29750f148e51aad30c/src/styles/slick-alpine-theme.scss#L279-L292)

The auto-height styles should also override the parent column:

```scss
.slick-header-column {
  height: auto;
  overflow: visible;
}
```

The current measurement also removes `--slick-auto-header-height`, leaving declarations such as:

```scss
height: calc(var(--slick-auto-header-height, auto) - 8px);
```

`calc(auto - 8px)` is invalid CSS. A safer pattern is:

```scss
.slick-header-column {
  height: auto;
  height: calc(var(--slick-auto-header-height) - 8px);
  overflow: visible;
}
```

For Alpine:

```scss
.slick-header-column {
  height: auto;
  height: calc(
    var(--slick-auto-header-height) -
    var(--alpine-auto-header-height-extra, v.$alpine-auto-header-height-extra)
  );
  overflow: visible;
}
```

The first `height: auto` declaration is used while the CSS variable is absent. Once JavaScript sets the variable, the calculated height takes effect.

The current clipping test does not fully verify this because it checks the first header column, which is the `#` column and is not a multi-line header.

### 2. Medium — programmatic autosizing does not remeasure header height

`reRenderColumns()` applies new header widths but does not call `recalculateHeaderHeight()`.

This matters for `IgnoreViewport` and `ignoreHeaderText`, because those settings can make columns narrower and cause their headers to wrap differently.

Suggested change:

```ts
reRenderColumns(reRender?: boolean) {
  this.applyColumnHeaderWidths();
  this.updateCanvasWidth(true);

  if (this._options.autoHeaderHeight) {
    this.recalculateHeaderHeight();
  }

  this.trigger(this.onAutosizeColumns, { columns: this.columns });

  if (reRender) {
    this.invalidateAllRows();
    this.render();
  }
}
```

The existing manual column resize path already recalculates at resize end, which is preferable to recalculating on every mouse-move.

### 3. Medium — container resizing needs to trigger recalculation

The current container resize handler only calls:

```ts
this.resizeCanvas.bind(this)
```

If the grid width changes responsively, header wrapping can change while the stored header height remains stale.

A minimal change would be:

```ts
this._bindingEventService.bind(this._container, 'resize', () => {
  if (this._options.autoHeaderHeight) {
    this.recalculateHeaderHeight();
  } else {
    this.resizeCanvas();
  }
});
```

If the application relies on ordinary CSS/container resizing, a `ResizeObserver` may be more reliable than a native `resize` event on the container. Any observer should include a width-change guard to avoid repeated layout loops.

### 4. Medium — frozen rows need explicit regression coverage

The original issue identified problems with both frozen columns and frozen rows:

- Different left/right header heights
- Grid overflow beyond its container
- Incorrect pane height calculations

The PR currently tests frozen columns, but not a frozen-row configuration.

At minimum, add a test using:

```ts
grid.setOptions({
  autoHeaderHeight: true,
  frozenColumn: 2,
  frozenRow: 5
});
```

The test should verify:

```ts
cy.get('.slick-header-left .slick-header-columns').then(($left) => {
  cy.get('.slick-header-right .slick-header-columns').should(($right) => {
    expect(Math.abs(
      $left[0].offsetHeight - $right[0].offsetHeight
    )).to.be.lessThan(2);
  });
});

cy.get('#myGrid').should(($grid) => {
  expect($grid[0].scrollHeight)
    .to.be.lte($grid[0].clientHeight + 1);
});
```

If this test fails, the fix belongs in `resizeCanvas()`. The enlarged column-header height must be included exactly once in the frozen top-pane/container calculation and must not be added again to the viewport height.

Because header height and available width can affect scrollbar presence, the recalculation should also use an only-if-changed guard or a bounded second pass.

### 5. Medium — the requested HTML/DOM scenarios are not committed as tests

The original issue explicitly mentions HTML strings and DOM elements. The PR discussion includes an additional example in a comment, but the committed example and Cypress spec do not currently test those cases.

Suggested test setup:

```ts
const headerElement = document.createElement('p');
headerElement.style.margin = '0';
headerElement.innerHTML = 'Customer <strong>Information</strong>';

const columns = [
  {
    id: 'html',
    name: 'Customer<br><strong>Information</strong>',
    field: 'title',
    width: 100
  },
  {
    id: 'dom',
    name: headerElement,
    field: 'title',
    width: 100
  }
];
```

The Cypress test should inspect every rendered header, not only the first column:

```ts
cy.get('.slick-header-column').each(($header) => {
  const header = $header[0] as HTMLElement;
  expect(header.scrollHeight).to.be.lte(header.clientHeight);
});
```

### 6. Minor — trailing whitespace

There is trailing whitespace at:

[[slick.grid.scss](https://github.com/6pac/SlickGrid/blob/e6548939c8bf336548adbf29750f148e51aad30c/src/styles/slick.grid.scss#L289)](https://github.com/6pac/SlickGrid/blob/e6548939c8bf336548adbf29750f148e51aad30c/src/styles/slick.grid.scss#L289)

## Verification

- Local checkout matches PR head `e6548939`.
- Local `npm run lint` passes.
- Local `npm run build:prod` passes.
- GitHub CI passes, including the full Cypress suite.
- All five inline review threads are resolved and outdated.
- All submitted reviews are `COMMENTED`; there is currently no approval.
- The latest top-level question from `muendlein` about `IgnoreViewport` and `ignoreHeaderText` remains unanswered.

## Overall recommendation

The PR has the right general direction and correctly keeps the feature opt-in. However, before merging, I recommend fixing the parent-column overflow and CSS fallback, adding recalculation after programmatic and responsive width changes, and adding regression coverage for HTML/DOM titles and frozen rows.

The current tests demonstrate the fixed example scenario, but they do not yet fully prove the requirements described in issue #1272.

File: src/styles/slick-alpine-theme.scss (hunk 2 above)
Comment (jahanbakhsh18) [reply src/styles/slick-alpine-theme.scss]: @ghiscoding Thank you again for your detailed feedback on the SASS variables. I really appreciate you taking the time... I wanted to follow up on the implementation to make sure we're on the same page.

I initially tried to use the pattern you suggested:
```scss
.slick-header-column {
  height: calc(
     var(--slick-auto-header-height, v.$alpine-auto-header-height) -
     var(--alpine-auto-header-height-extra, v.$alpine-auto-header-height-extra)
  );
```
However, during testing, I found that when `--slick-auto-header-height` is set by JavaScript (e.g., to 54px), the calc() falls back to the SASS value instead of using the CSS variable. After some investigation, I believe this happens because `calc()` is sensitive to how the CSS variable is resolved.

The pattern that consistently worked for me is the double declaration:
```scss
.slick-header-column {
  height: auto;
  height: calc(
    var(--slick-auto-header-height) -
    var(--alpine-auto-header-height-extra, v.$alpine-auto-header-height-extra)
  );
  overflow: visible;
}
```

Does this approach align with your expectations? I'm happy to adjust if you'd prefer a different pattern. I just wanted to make sure the code is reliable across all browsers.

Thanks again for your guidance!

Comment (ghiscoding) [reply src/styles/slick-alpine-theme.scss]: Well the way I see is that SASS is used at compiled time, but CSS variables are used in runtime. Also again I took advantage of the CSS variable fallback because it allows for scenario that if no CSS var is provided then use the SASS definition, but CSS variables should always win when provided at runtime, at least that's my believe. But I'm not sure how it all plays out when using calc which happens at runtime and I don't see anything special in the [CSS Variables](https://developer.mozilla.org/en-US/docs/Web/CSS/Guides/Cascading_variables/Using_custom_properties#custom_property_fallback_values) on MDN... but the Codex review that I posted at the bottom does hint at an error in your calc() usage, which might explain why it doesn't behave like you expect (also with your code in calc, you have no fallback and that could be problematic if for example the CSS var is null then your calc would totally fail)... I would suggest you review the entire Codex [review](https://github.com/6pac/SlickGrid/pull/1282#issuecomment-5365004544) below, I asked it to also provide code suggestions when possible, so it should be useful.

**Comment by jahanbakhsh18** [comment]
@ghiscoding @muendlein  Thank you both for the detailed feedback, and apologies for the delayed response... 
(I was away for a few days).

I’ve spent some time going through the interactions between `autoHeaderHeight`, the different column autosizing modes, frozen columns/rows, and the header sizing/recalculation flow. There are a few more interactions here than I initially expected, so I wanted to understand the behavior properly rather than make isolated changes based only on the individual review comments.

Regarding `IgnoreViewport` / `ignoreHeaderText`, I’m now adding explicit coverage for those scenarios. I also found that it makes more sense to separate the example into two focused examples:

- one focused on the `auto-header-height mechanism` itself, including enable/disable, frozen columns/rows, and keeping the header panes aligned
- another focused on `different header/content types and column autosizing`, including HTML strings, DOM elements, regular text, `IgnoreViewport`, `ignoreHeaderText`, and related `autosizing` behavior.

This should make the examples and regression tests much easier to understand and should also directly cover the scenario you mentioned, @muendlein.

I’ll continue with the remaining review points once these interactions are settled. Thanks again for the feedback and for your patience.

**Comment by jahanbakhsh18** [comment]
@6pac, @ghiscoding, @muendlein please accept my apologies for the delay on this PR.

While testing **Review 2: “Programmatic autosizing does not remeasure header height”**, I noticed that simply adding `recalculateHeaderHeight()` to `reRenderColumns()` causes a recursive loop and stack overflow with the `LegacyForceFit` autosize mode.
Because of that, I decided to test all six autosize modes more thoroughly and include them in the second example. This helped me identify and address a few related edge cases, particularly around container resizing and header height recalculation.

I’m working through the remaining reviews now and will do my best to finalize and cover all six reviews ASAP. Thank you for your patience!

**Comment by ghiscoding** [comment]
oops my latest PRs caused a single line conflict, shouldn't be too hard to fix though which seems to be to keep both lines.

**Comment by jahanbakhsh18** [comment]
Hi everyone,
I’ve now addressed the review feedback and completed the implementation for the auto header height changes. 

### What was updated
* Fixed header content clipping for both **Classic** and **Alpine** themes.
* Ensured `.slick-header-column` allows the automatically calculated header height and rendered HTML/DOM content to be visible.
* Added the necessary CSS fallback for the auto header height variable.
* Added `recalculateHeaderHeight()` to the `reRenderColumns()` pipeline so programmatic column autosizing correctly updates the header height.
* Verified the behavior with autosizing modes
* Investigated the `LegacyForceFit` resize pipeline and avoided introducing the recursive loop that could cause a stack overflow.
* Kept the existing container resize pipeline unchanged for the other autosize modes. Their recalculation can be triggered through **Apply**, while `LegacyForceFit` continues to react through the existing resize pipeline.
* Added regression coverage for:
  * Multi-line HTML headers
  * DOM element headers
  * Frozen columns and frozen rows
  * Header height toggling
  * Column resizing
  * `ignoreHeaderText`
  * Container resizing
  * Recursive rendering / stack-overflow scenarios

I also added a second example specifically for demonstrating the interaction between **Auto Header Height** and the different column autosizing modes.

#### Demo: Auto Header Height and Autosize Modes


https://github.com/user-attachments/assets/f64dc5e8-045e-461d-bf13-1abed1bccaa7



Thanks again for the reviews and guidance. I believe the current implementation and test coverage now provide a good regression suite for the original issue.

**Comment by ghiscoding** [comment]
@jahanbakhsh18 I've asked Codex for another review and it still identified some small change requests, could you please look into this? If you need help, I can always ask Codex to fix it, let me know if you need help or not. It would be great to have this released this week. Thanks

Here's the Codex review with code suggestions separated in 2 comments:

Thanks for the work on this — the feature direction looks good. I think we need one implementation update and stronger regression coverage before merging.

### Programmatic autosizing leaves auto header height stale

`reRenderColumns()` updates the header widths, but it does not remeasure `autoHeaderHeight`.

This matters for the `IgnoreViewport` + `ignoreHeaderText` case mentioned above: autosizing can make a column narrower than its header text requires, causing the header to wrap after its height has already been fixed. The old CSS variable/header height then remains in place, so the newly wrapped title can be clipped.

Suggested change:

```ts
reRenderColumns(reRender?: boolean) {
  this.applyColumnHeaderWidths();
  this.updateCanvasWidth(true);

  if (this._options.autoHeaderHeight) {
    this.recalculateHeaderHeight();
  }

  this.trigger(this.onAutosizeColumns, { columns: this.columns });

  if (reRender) {
    this.invalidateAllRows();
    this.render();
  }
}
```

**Comment by ghiscoding** [comment]
Part 2 is about adding more Cypress E2E test coverage:

Please add explicit Cypress coverage for the missing scenarios. For example:

```ts
it('re-measures after IgnoreViewport autosizing ignores header text', () => {
  let initialHeaderHeight = 0;

  cy.window().then((win) => {
    const grid = (win as any).grid;
    const durationColumn = grid.getColumns().find((column: any) => column.id === 'duration');

    grid.setOptions({
      autosizeColsMode: (win as any).Slick.GridAutosizeColsMode.IgnoreViewport,
    });

    durationColumn.name = 'Very long duration header that should wrap after autosizing';
    durationColumn.width = 260;
    durationColumn.autoSize = {
      ...durationColumn.autoSize,
      ignoreHeaderText: true,
    };

    grid.setColumns(grid.getColumns());
  });

  cy.get('.slick-header-columns').then(($headers) => {
    initialHeaderHeight = $headers[0].offsetHeight;
  });

  cy.window().then((win) => {
    (win as any).grid.autosizeColumns();
  });

  cy.get('.slick-header-column[data-id="duration"]').should(($header) => {
    const header = $header[0] as HTMLElement;
    expect(header.scrollHeight).to.be.lte(header.clientHeight);
  });

  cy.get('.slick-header-columns').should(($headers) => {
    expect($headers[0].offsetHeight).to.be.greaterThan(initialHeaderHeight);
  });
});
```

This should fail against the current implementation: autosizing narrows the column and wraps the header, but `reRenderColumns()` does not call `recalculateHeaderHeight()`.

Please also add the HTML and DOM title cases to the committed example:

```js
const headerElement = document.createElement('p');
headerElement.style.margin = '0';
headerElement.innerHTML = 'Customer <strong>Information</strong>';

{
  id: 'html-header',
  name: 'Customer<br><strong>Information</strong>',
  field: 'title',
  width: 100,
},
{
  id: 'dom-header',
  name: headerElement,
  field: 'title',
  width: 100,
},
```

…and cover them with a test that checks the actual headers rather than the first `#` column:

```ts
it('does not clip HTML or DOM header content', () => {
  cy.get(
    '.slick-header-column[data-id="html-header"], ' +
    '.slick-header-column[data-id="dom-header"]'
  ).each(($header) => {
    const header = $header[0] as HTMLElement;
    expect(header.scrollHeight).to.be.lte(header.clientHeight);
  });
});
```

**Comment by jahanbakhsh18** [comment]
Thanks @ghiscoding!

Could you please check the latest commit of the PR once more? I believe some of these comments may be based on an earlier version.

The latest version already includes:
- The recalculateHeaderHeight() call in reRenderColumns().
- The new example-auto-header-height-autosize.html example covering the autosizing modes, including IgnoreViewport and ignoreHeaderText.
- The HTML and DOM header scenarios in the new example.
- The corresponding Cypress coverage, including the 18 tests in example-auto-header-height-autosize.cy.ts.

I’ll be happy to add anything that is still missing after reviewing the latest version 🙂

**Comment by ghiscoding** [comment]
oops you're right, my bad, I did a git checkout of your branch again to do the review but forgot to git pull recent changes, dohhhh 🤦🏻‍♂️🤣  doing another review now

**Comment by ghiscoding** [comment]
ok it replied with: 

Yes — after updating to `385258966`, I consider the PR mergeable as-is, assuming normal CI is green.

The author addressed the earlier blockers:

- `reRenderColumns()` now remeasures when `autoHeaderHeight` is on, with a height-change guard to keep `LegacyForceFit` recursion bounded. [slick.grid.ts](/home/ghislainb/sourcecode/SlickGrid/src/slick.grid.ts:1995) and [slick.grid.ts](/home/ghislainb/sourcecode/SlickGrid/src/slick.grid.ts:3341)
- The CSS now uses the safe `height: auto` fallback and enables parent-header overflow for auto-height mode. [slick.grid.scss](/home/ghislainb/sourcecode/SlickGrid/src/styles/slick.grid.scss:277)
- The new autosize example includes HTML and DOM titles plus `IgnoreViewport`/`ignoreHeaderText`. [example-auto-header-height-autosize.html](/home/ghislainb/sourcecode/SlickGrid/examples/example-auto-header-height-autosize.html:110)
- The new 18-test Cypress spec checks actual HTML/DOM headers and all rendered headers after autosizing. [example-auto-header-height-autosize.cy.ts](/home/ghislainb/sourcecode/SlickGrid/cypress/e2e/example-auto-header-height-autosize.cy.ts:15)
- Frozen rows/columns are now covered in the existing feature spec. [example-auto-header-height.cy.ts](/home/ghislainb/sourcecode/SlickGrid/cypress/e2e/example-auto-header-height.cy.ts:54)

I found no remaining merge-blocking issue and have no required code suggestions. `git diff --check` is clean; the production build and lint passed on this exact commit.

... so it looks good and ready to merge, I'll do that later today and maybe go with a new release too

File: src/slick.grid.ts [hunk 6]
```diff
@@ -1962,6 +1968,78 @@ export class SlickGrid<TData = any, C extends Column<TData> = Column<TData>, O e
...
+    if (maxHeight <= 0) {
+      return;
+    }
+
+    this._setAutoHeaderHeightStyles(maxHeight, headers);
+
+    // Only resize if the height actually changed
+    if (Math.abs(maxHeight - currentHeight) > 0.5) {
+      this.resizeCanvas();
+    }
```
Comment (ghiscoding) [inline src/slick.grid.ts]:
```suggestion
    if (maxHeight > 0) {
      this.setAutoHeaderHeightStyles(maxHeight, headers);

      // A viewport resize is only necessary when the calculated height actually changed.
      if (Math.abs(maxHeight - currentHeight) > 0.5) {
        this.resizeCanvas();
      }
    }
```

Comment (ghiscoding) [reply src/slick.grid.ts]: oops I made a typo in my suggestion, fixed in a follow-up commit. Anyway, the idea was to simply decrease lines of code and make it more readable

**Comment by 6pac** [comment]
If you guys are happy with it, go ahead

**Comment by ghiscoding** [comment]
Thanks again for your contribution, I'll push a new release soon :)
See [v5.20.0](https://github.com/6pac/SlickGrid/releases/tag/5.20.0)
